### ISSUE TITLE
Data loss: troublesome interaction with other cloud file services

### ISSUE THREAD

**Issue by despens** [nextcloud/desktop#4276]
## Different cloud file sharing services fight for control over files

I provided nextcloud services for a few ad-hoc groups that needed to collaborate for a limited time on a project. These groups were composed of freelancers, a mix of Windows, Macintosh, and Linux systems, and because of the need to work with different customers have almost every possible cloud file sharing software on their computers. Nextcloud does not play well with them, in particular the "default" ones for Windows and Apple, OneDrive and iCloud.

When Mac users use iCloud to "optimize storage" on their computer, iCloud will remove what it considers old files from local storage and move them to iCloud, without the user noticing. Nextcloud regards such files as deleted, and dutifully deletes them for collaborators that users have shared their directory with. — This has led to data loss, on older files, that is not easy to notice immediately.

When Windows users use nextcloud on a directory that is also managed by OneDrive, it can happen that files are deleted for everyone else that is using the shared directory, but remain visible for the Windows user, only a few minutes after the files have been created or moved into the directory. (I believe OneDrive uses some kind of overlay or virtual file system that again is inaccessible to nextcloud's desktop client.) — Again, this has led to data loss, files "blinking" in and out of existence on users' disks, and gigantic modification/deletion logs on the nextcloud server, as affected files were "deleted" on a Windows machine, synced back to nextcloud from a Mac, being deleted by Windows again, and so forth.

The design of commercial cloud file services, which aim to appear as basic infrastructure rather than a choice, makes debugging of such issues incredibly hard. Many users do not even know they have iCloud activated, or what directories iCloud is managing for them.

## What nextcloud could do

It might be impossible for the nextcloud desktop client to adapt to all the ways commercial cloud file sharing vendors pretend that a file exists locally, but perhaps it could be checked if
- a certain directory is already managed by another file sharing service,
- iCloud's storage management is activated for a complete computer system or certain directories, or
- nextcloud is configured to managed any kind of "overlay file system" or remotely mounted file system that on network disconnection could appear as empty to the nextcloud client

and display a warning to users that there is a risk of setting this up, or mark affected directories as problematic in the "settings" view with some icon.

I hope this could create more trust in nextcloud on the side of users (who of course believe this weird behavior is the fault of nextcloud, since everything worked fine before) and significantly reduce debugging time.

**Comment by despens** [comment]
Other file sharing/backup vendors alerting users about this this problem:
- [CarbonCloner](https://bombich.com/kb/ccc6/limitations-online-only-placeholder-files)

**Comment by despens** [comment]
Hey I'm coming back to prevent any bots from setting this issue to "stale" or whatever. Let me stress that this is a huge issue outside of controlled corporate environments. It is causing data loss and I will definitely not be able to help with debugging because everyone who agreed to use nextcloud on my recommendation experienced data loss that took days to recover from other backups—these folks will not serve as test candidates once more.

**Comment by joshtrichards** [comment]
https://docs.nextcloud.com/desktop/3.13/troubleshooting.html#isolating-other-issues

**Comment by despens** [comment]
@joshtrichards I agree this is challenging, maybe before getting to the bottom of the technical aspects, it would make sense to include a message into the sync client that warns users to be cautious about which directories to sync. With the introduction of Windows 11, many users are reporting that OneDrive was managing directories without them even realizing. Users shouldn't need to look for trouble shooting information after they lost data, but be warned of possible pitfalls beforehand.

**Comment by mophilly** [comment]
The language used in the troubleshooting section is technically correct. It is written at a post-graduate reading level, which is challenging. One change that could help is to add a list of software known to cause problems. Using specific names will be helpful to those new to these types of systems and to those who are in a hurry. 

For example, the current text:
- If you find that the results of the synchronizations are unreliable, please ensure that the folder to which you are synchronizing is not shared with other synchronization applications.

- Synchronizing the same directory with Nextcloud and other synchronization software such as Unison, rsync, Microsoft Windows Offline Folders, or other cloud services such as Dropbox or Microsoft SkyDrive is not supported and should not be attempted. In the worst case, it is possible that synchronizing folders or files using Nextcloud and other synchronization software or services can result in data loss.

Rewriting these two paragraphs may help. To wit:
- Check that the NextCloud sync folder is not in a directory managed by another synchronization software system, such as  DropBox, Google Drive, iCloud, Microsoft SkyDrive, or similar systems. Missing files, incomplete synchronization, and other problems will arise when a sync folder is managed by more than one sync system. In the worst case, synchronizing folders or files using Nextcloud and another synchronization software or service may result in data loss.

**Comment by mophilly** [comment]
> This information needs to go into the sync client. Folks consult a trouble shooting guide when a problem arises, and at that point the data is likely to be lost already.

Excellent point. Place the text on the client panel where the user selects the local target directory.

**Comment by Rello** [comment]
Hello,

no sync clients - no matter which provider - should be nested or combined. 
with macOS VFS, this is also not possible anymore as it will always be parallel to other Apple File Provider extensions.
We do not have a technical chance to detect overlapping setups with other sync clients, backups, roaming profiles or so. This has to be in the responsibility of the user to use all tools correclty

**Comment by feld** [comment]
On MacOS you could

- add `.nosync` to the end of the directory name; provide a symlink using the original directory name to make this transparent
- add the `isExcludedFromBackupKey` property to the directory so MacOS refuses to do anything with them, including moving them to iCloud
- blame the users for holding it wrong
